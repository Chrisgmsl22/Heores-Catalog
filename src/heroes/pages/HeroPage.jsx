import React, { useMemo } from 'react'
import { Navigate, useParams } from 'react-router-dom';
import { getHeroById } from '../helpers/getHeroById';

export const HeroPage = () => {

    const {id} = useParams();
    const currentHero = useMemo(() => getHeroById(id), [id]); 
    console.log(currentHero);

    const onNavigateBack = () => {
        window.history.back();
    }

    if (!currentHero) {
        return <Navigate to='/marvel'/>
    }
    return (
        <div className='row mt-5'>
            <div className="col-4">
                <img
                    src={`../assets/heroes/${id}.jpg`}
                    alt={currentHero.superhero}
                    className='img-thumbnail animate__animated animate__fadeInLeft'
                />
            </div>

            <div className="col-8">
                <h3>{currentHero.superhero}</h3>
                <ul className="list-group list-group-flush">
                    <li className="list-group-item">
                        <b>Alter ego:</b> {currentHero.alter_ego}
                    </li>
                    <li className="list-group-item">
                        <b>Publisher:</b> {currentHero.publisher}
                    </li>
                    <li className="list-group-item">
                        <b>First appearance:</b> {currentHero.first_appearance}
                    </li>
                </ul>

                <h5 className='mt-3'>Characters</h5>
                <p>{currentHero.characters}</p>

                <button 
                    className='btn btn-outline-info'
                    onClick={onNavigateBack}
                >
                    Return
                </button>
            </div>
        </div>

    )
}


// () => window.history.back() para ir hacia atras