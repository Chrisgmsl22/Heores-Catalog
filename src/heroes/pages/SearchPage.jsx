import React from 'react'
import { HeroCard } from '../components/HeroCard'
import { useForm } from '../../hooks/useForm'
import { useLocation, useNavigate } from 'react-router-dom';
import queryString from 'query-string';
import { getHeroesByName } from '../helpers/getHeroesByName';

export const SearchPage = () => {

  // React router dom hooks
  const navigate = useNavigate();
  const location = useLocation(); // Nos da la informacion de la ruta actual

  // Extraemos el query string
  const {q = '' } = queryString.parse(location.search);
  const heroes = getHeroesByName(q);
  const showSearch = (q.length === 0) ? true : false;
  const showError = (q.length > 0) && (heroes.length === 0) ? true : false;

  const {searchText, onInputChange} = useForm({
    searchText: q,
  });

  const handleSearchSubmit = (e) => {
    e.preventDefault(); //Evita que se recargue la pagina
    
    // hacemos la navegacion a la ruta de search
    navigate(`/search?q=${searchText.toLowerCase().trim()}`);

  }


  return (
    <>
      <h1>Search</h1>
      <hr />

      <div className="row">
        <div className="col-5">
          <h4>Searching</h4>
          <hr />

          <form onSubmit={handleSearchSubmit} aria-label='form'>
            <input
              type="text"
              placeholder="Find your hero"
              className="form-control"
              name="searchText"
              autoComplete="off"
              value={searchText}
              onChange={onInputChange}
            />
            <button
              type="submit"
              className="btn mt-1 btn-block btn-outline-secondary"
            >
              Search...
            </button>
          </form>
        </div>

        <div className="col-7">
          <h4>Results</h4>
          <hr />

          {/* {
            (q === '') 
            ?  <div className="alert alert-info"> Search a hero </div>
            : (heroes.length === 0) && <div className="alert alert-danger"> There's no response on <b>{q}</b> </div>
              
          } */}

          <div 
            className="alert alert-info animate__animated animate__fadeIn"
            style={{display: (showSearch) ? '' : 'none'}}
          >
            Search a hero
          </div>
          <div 
            className="alert alert-danger animate__animated animate__fadeIn"
            style={{display: (showError) ? 'block' : 'none'}}
            aria-label='error'
          >
            There's no response on <b>{q}</b>
          </div>

          {
            heroes.map(hero => (
              <HeroCard
                key={hero.id}
                {...hero}
              />
            ))
          }
        </div>
      </div>
    </>
  )
}
