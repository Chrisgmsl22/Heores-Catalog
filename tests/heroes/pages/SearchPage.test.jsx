import { fireEvent, render, screen } from "@testing-library/react";
import { SearchPage } from "../../../src/heroes/pages/SearchPage";
import { MemoryRouter, useNavigate } from "react-router-dom";

// requireActual para el useNavigate
const mockedUseNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockedUseNavigate,
}));


describe('Pruebas en el componente SearchPage', () => {

    beforeAll(() => {
        jest.clearAllMocks();
    });

    test('Debe de mostrarse correctamente con valores por defecto', () => {
        const {container} = render(
            <MemoryRouter>
                <SearchPage />
            </MemoryRouter>
        );

        // screen.debug();

        expect(container).toMatchSnapshot();
    });

    test('Debe de mostrar a Batman y el input con el valor del QueryString', () => {
        const {container} = render(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <SearchPage />
            </MemoryRouter>
        );
        const input = screen.getByRole('textbox');
        const title = screen.getByText('Batman');
        const img = screen.getByRole('img');

        //screen.debug();

        //expect(screen.getByRole('textbox')).toHaveValue('batman');
        expect(input.value).toBe('batman');
        expect(title).toBeTruthy();
        expect(img.src).toContain('/heroes/dc-batman.jpg');
        expect(container).toMatchSnapshot();
    });

    test('Debe de mostrar error si no se encuentra un hÃ©roe', () => {
        const {container} = render(
            <MemoryRouter initialEntries={['/search?q=batman123']}>
                <SearchPage />
            </MemoryRouter>
        );
        // Obtenemos del aria-label
        const error = screen.getByLabelText('error');

        // screen.debug();

        expect(error).toBeTruthy();
        expect(error.style.display).toBe('block');
        expect(container).toMatchSnapshot();
        
    });

    test('Debe de llamar el navigate a la pantalla nueva', () => {
        render(
            <MemoryRouter initialEntries={['/search']}>
                <SearchPage />
            </MemoryRouter>
        );
        screen.debug();
        
        const input = screen.getByRole('textbox');
        fireEvent.change(input, {target: {value: 'spiderman'}});

        const form = screen.getByRole('form');
        fireEvent.submit(form);

        expect(mockedUseNavigate).toHaveBeenCalled();
        expect(mockedUseNavigate).toHaveBeenCalledWith('/search?q=spiderman');
    });
    
    
    

    
    
});
