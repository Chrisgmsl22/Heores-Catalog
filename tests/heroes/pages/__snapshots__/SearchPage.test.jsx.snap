// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente SearchPage Debe de mostrar a Batman y el input con el valor del QueryString 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-5"
    >
      <h4>
        Searching
      </h4>
      <hr />
      <form
        aria-label="form"
      >
        <input
          autocomplete="off"
          class="form-control"
          name="searchText"
          placeholder="Find your hero"
          type="text"
          value="batman"
        />
        <button
          class="btn mt-1 btn-block btn-outline-secondary"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      class="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        class="alert alert-info animate__animated animate__fadeIn"
        style="display: none;"
      >
        Search a hero
      </div>
      <div
        aria-label="error"
        class="alert alert-danger animate__animated animate__fadeIn"
        style="display: none;"
      >
        There's no response on 
        <b>
          batman
        </b>
      </div>
      <div
        class="col animate__animated animate__fadeIn"
      >
        <div
          class="card"
        >
          <div
            class="row no-gutters"
          >
            <div
              class="col-4 "
            >
              <img
                alt="Batman"
                class="card-img"
                src="/heroes/dc-batman.jpg"
              />
            </div>
            <div
              class="col-8"
            >
              <div
                class="card-body"
              >
                <h5
                  class="card-title"
                >
                  Batman
                </h5>
                <p
                  class="card-text"
                >
                  Bruce Wayne
                </p>
                <p
                  class="card-text"
                />
                <p
                  class="card-text"
                >
                  <small
                    class="text-muted"
                  >
                              
                    Detective Comics #27
                  </small>
                </p>
                <a
                  href="/hero/dc-batman"
                >
                  Mas...
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas en el componente SearchPage Debe de mostrar error si no se encuentra un h√©roe 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-5"
    >
      <h4>
        Searching
      </h4>
      <hr />
      <form
        aria-label="form"
      >
        <input
          autocomplete="off"
          class="form-control"
          name="searchText"
          placeholder="Find your hero"
          type="text"
          value="batman123"
        />
        <button
          class="btn mt-1 btn-block btn-outline-secondary"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      class="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        class="alert alert-info animate__animated animate__fadeIn"
        style="display: none;"
      >
        Search a hero
      </div>
      <div
        aria-label="error"
        class="alert alert-danger animate__animated animate__fadeIn"
        style="display: block;"
      >
        There's no response on 
        <b>
          batman123
        </b>
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas en el componente SearchPage Debe de mostrarse correctamente con valores por defecto 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-5"
    >
      <h4>
        Searching
      </h4>
      <hr />
      <form
        aria-label="form"
      >
        <input
          autocomplete="off"
          class="form-control"
          name="searchText"
          placeholder="Find your hero"
          type="text"
          value=""
        />
        <button
          class="btn mt-1 btn-block btn-outline-secondary"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      class="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        class="alert alert-info animate__animated animate__fadeIn"
      >
        Search a hero
      </div>
      <div
        aria-label="error"
        class="alert alert-danger animate__animated animate__fadeIn"
        style="display: none;"
      >
        There's no response on 
        <b />
      </div>
    </div>
  </div>
</div>
`;
